### PR TITLE
TS-4751: Prune cached headers before merging the updated response.

### PR DESCRIPTION
**Pull request by jpeach**
When we update a cached response with a new server response, make
sure to delete any cache-related beforehand. This is needed to
handle the situation where an origin returns Age on an initial
response but not on subsequent responses. We should not preserve
the initial Age in this case.

### CODE REVIEW

**Comment by jpeach** [comment]
Ping @zwoop @bryancall @SolidWallOfCode

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/531/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/428/ for details.

File: proxy/http/HttpTransact.cc [hunk 1]
```diff
@@ -4785,27 +4785,34 @@ HttpTransact::merge_and_update_headers_for_cache_update(State *s)
...
   }
 
-  merge_response_header_with_cached_header(s->cache_info.object_store.response_get(), &s->hdr_info.server_response);
+  // Delete caching headers from the cached response. If these are still being served
+  // by the origin we will copy new versions in from the server response.
+  cached_hdr->field_delete(MIME_FIELD_AGE, MIME_LEN_AGE);
+  cached_hdr->field_delete(MIME_FIELD_CACHE_CONTROL, MIME_LEN_CACHE_CONTROL);
+  cached_hdr->field_delete(MIME_FIELD_ETAG, MIME_LEN_ETAG);
+  cached_hdr->field_delete(MIME_FIELD_EXPIRES, MIME_LEN_EXPIRES);
+  cached_hdr->field_delete(MIME_FIELD_LAST_MODIFIED, MIME_LEN_LAST_MODIFIED);
```
Comment (bryancall) [inline proxy/http/HttpTransact.cc]: I don't think it is a good idea to delete Expires, Last-Modifed, or Cache-Control from the cached object.  They are not required to be sent from the origin according to the HTTP/1.1 RFC on a 304:
https://tools.ietf.org/html/rfc2616#section-10.3.5

The newer RFC (https://tools.ietf.org/html/rfc7232#section-4.1) changed this, but we should still support RFC2616.

**Comment by jpeach** [comment]
I guess the key phrase from RFC 2616 is ``if the field-value might differ from that sent in any previous response``. What happens if the origin had an ``Expires`` but not it doesn't? RFC7232 would say that we MUST not use the ``Expires`` we have, but RFC2616 says we should. I think this could easily happen with an origin server configuration change.

**Comment by zwoop** [comment]
The safer of the two seems to be to honor the last seen header.

**Comment by jpeach** [comment]
The problem we hit was we had a bogus Age (12000) header cached. Since the origin was no linger sending an Age header, the object would never become fresh.

One common case could be an origin that sends Expires, then is reconfigured to not send Expires. In this case, we would always keep the Expires from the cached header which would be wrong.

**Comment by jpeach** [comment]
AFAICT the RFC7232 language was introduced in HTTPbis ticket [#406](https://tools.ietf.org/wg/httpbis/trac/ticket/406). It seems like this change was regarded as a bug fix to RFC2616 rather than a change in requirement or semantics.

**Comment by bryancall** [comment]
@jpeach I would be OK with deleting the Expires header, but I think we should keep the Cache-Control, Last-Modified, and Vary.

**Comment by jpeach** [comment]
Updated to only delete Age, ETag and Expires from the cached header.

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/543/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/544/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/440/ for details.

**Comment by bryancall** [comment]
👍 - Looks good

**Pull request merged**